File: LHDS.AdminPortal.Web/ClientApp/src/brokers/apiBroker.datasetspecifications.ts
import { Guid } from "guid-typescript";
import { DataSetSpecification } from "../models/dataSetSpecifications/dataSetSpecification";
import ApiBroker from "./apiBroker";

class DataSetSpecificationBroker {
    relativeDataSetSpecificationUrl = '/api/DataSetSpecifications';

    private apiBroker: ApiBroker = new ApiBroker();

    async PostDataSetSpecificationAsync(dataSetSpecification: DataSetSpecification) {
        return await this.apiBroker.PostAsync(this.relativeDataSetSpecificationUrl, dataSetSpecification)
            .then(result => new DataSetSpecification(result.data));
    }

    async GetAllDataSetSpecificationsAsync(queryString: string) {
        const url = this.relativeDataSetSpecificationUrl + queryString;
        
        if (queryString === "/") {
            return undefined;
        }
        
        return await this.apiBroker.GetAsync(url)
            .then(result => result.data.map((optOut: any) => new DataSetSpecification(optOut)));
    }
}

export default DataSetSpecificationBroker;

File: LHDS.AdminPortal.Api.Tests.Acceptance/Brokers/ApiBroker.DataSetSpecifications.cs
// ---------------------------------------------------------------
// Copyright (c) North East London ICB. All rights reserved.
// ---------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using LHDS.AdminPortal.Api.Tests.Acceptance.Models.DataSetSpecifications;
using LHDS.AdminPortal.Api.Tests.Acceptance.Models.OdataResponses;

namespace LHDS.AdminPortal.Api.Tests.Acceptance.Brokers
{
    public partial class ApiBroker
    {
        private const string dataSetSpecificationsRelativeUrl = "api/dataSetSpecifications";
        private const string dataSetSpecificationsRelativeOdataUrl = "odata/dataSetSpecifications";

        public async ValueTask<DataSetSpecification> PostDataSetSpecificationAsync(
            DataSetSpecification dataSetSpecification) =>
                await this.apiFactoryClient.PostContentAsync(dataSetSpecificationsRelativeUrl, dataSetSpecification);

        public async ValueTask<List<DataSetSpecification>> GetAllDataSetSpecificationsAsync()
        {
            OdataResponse<DataSetSpecification> response =
                await this.apiFactoryClient.GetContentAsync<OdataResponse<DataSetSpecification>>(
                    $"{dataSetSpecificationsRelativeOdataUrl}/");

            return response.Items;
        }

        public async ValueTask<DataSetSpecification> GetDataSetSpecificationByIdAsync(Guid dataSetSpecificationId) =>
            await this.apiFactoryClient.GetContentAsync<DataSetSpecification>(
                $"{dataSetSpecificationsRelativeUrl}/{dataSetSpecificationId}");

        public async ValueTask<DataSetSpecification> PutDataSetSpecificationAsync(
            DataSetSpecification dataSetSpecification) =>
                await this.apiFactoryClient.PutContentAsync(dataSetSpecificationsRelativeUrl, dataSetSpecification);

        public async ValueTask<DataSetSpecification> DeleteDataSetSpecificationByIdAsync(Guid dataSetSpecificationId) =>
            await this.apiFactoryClient.DeleteContentAsync<DataSetSpecification>(
                $"{dataSetSpecificationsRelativeUrl}/{dataSetSpecificationId}");
    }
}
